clients/drcachesim/tests/scheduler_unit_tests.cpp

```diff
@@ -6908,6 +6909,101 @@ test_options_match()
+    generate_trace_entry() override
+    {
+        trace_entry_t generated_entry = { TRACE_TYPE_READ, 4, { addr_to_generate_ } };
+        return generated_entry;
+    }
+
+private:
+    addr_t addr_to_generate_ = 0x0;
+};
+
+static void
+test_noise_generator()
+{
+    std::cerr << "\n----------------\nTesting noise generator\n";
+    static constexpr addr_t noise_generator_addr_to_generate = 0xdeadbeef;
+    static constexpr memref_tid_t TID_A = 42;
+    static constexpr memref_tid_t TID_B = 99;
+    std::vector<trace_entry_t> refs_A = {
+        /* clang-format off */
+        make_thread(TID_A),
+        make_pid(1),
+        make_version(4),
+        make_timestamp(10),
+        make_instr(10),
+        make_timestamp(30),
+        make_instr(30),
+        make_timestamp(50),
+        make_instr(50),
+        make_exit(TID_A),
+        /* clang-format on */
+    };
+    std::vector<trace_entry_t> refs_B = {
+        /* clang-format off */
+        make_thread(TID_B),
+        make_pid(1),
+        make_version(4),
+        make_timestamp(20),
+        make_instr(20),
+        make_timestamp(40),
+        make_instr(40),
+        make_timestamp(60),
+        make_instr(60),
+        make_exit(TID_B),
+        /* clang-format on */
+    };
+    std::vector<scheduler_t::input_reader_t> readers;
+    readers.emplace_back(std::unique_ptr<mock_reader_t>(new mock_reader_t(refs_A)),
+                         std::unique_ptr<mock_reader_t>(new mock_reader_t()), TID_A);
+    readers.emplace_back(std::unique_ptr<mock_reader_t>(new mock_reader_t(refs_B)),
+                         std::unique_ptr<mock_reader_t>(new mock_reader_t()), TID_B);
+    // Add noise.
+    noise_generator_info_t noise_generator_info = { 1, 1, 10 };
+    readers.emplace_back(
+        std::unique_ptr<mock_noise_generator_t>(new mock_noise_generator_t(
+            noise_generator_info, noise_generator_addr_to_generate)),
+        std::unique_ptr<mock_noise_generator_t>(new mock_noise_generator_t()),
+        INVALID_THREAD_ID);
+    scheduler_t scheduler;
+    std::vector<scheduler_t::input_workload_t> sched_inputs;
+    sched_inputs.emplace_back(std::move(readers));
+    if (scheduler.init(sched_inputs, 1,
+                       scheduler_t::make_scheduler_serial_options(/*verbosity=*/4)) !=
+        scheduler_t::STATUS_SUCCESS)
+        assert(false);
+    auto *stream = scheduler.get_stream(0);
+    memref_t memref;
+    bool found_at_least_one_read = false;
+    for (scheduler_t::stream_status_t status = stream->next_record(memref);
+         status != scheduler_t::STATUS_EOF; status = stream->next_record(memref)) {
+        assert(status == scheduler_t::STATUS_OK);
+        // There is just one output so we expect to always see 0 as the ordinal.
+        assert(stream->get_input_workload_ordinal() == 0);
+        // All TRACE_TYPE_READ records are generated by the noise generator and their
+        // addr is always the constant noise_generator_addr_to_generate.
```

Comment:
You don't need to do this arrangement with the only reads in the noise: you have the tid of the noise input, and you also have the ordinal of the noise input.  You can use one of those, and then put reads in the other inputs, which seems a stronger check.

Reply:
Done.
Now using the input ordinal to identify the noise generator.
Also added reads to the other input workloads.